File: 1268-search-suggestions-system/1268-search-suggestions-system.cpp
struct TrieNode {
    vector<TrieNode*> child{ vector<TrieNode*>(26, nullptr) };
    string* word{ };
};

void insert(TrieNode* root, string& word) {
    for (char c : word) {
        c -= 'a';
        if (root->child[c] == nullptr) {
            root->child[c] = new TrieNode();
        }
        root = root->child[c];
    }
    root->word = &word;
}

void dfs(TrieNode* root, int limit, vector<string>& result) {
    if (root == nullptr) {
        return;
    }
    if (result.size() == limit) {
        return;
    }
    if (root->word != nullptr) {
        result.push_back(*root->word);
    }
    for (int i = 0; i < 26; i++) {
        dfs(root->child[i], limit, result);
    }
}

vector<string> getWords(TrieNode* root, int limit) {
    vector<string> result;
    dfs(root, limit, result);
    return result;
}

class Solution {
public:

    vector<vector<string>> suggestedProducts(vector<string>& products, string searchWord) {
        TrieNode* root = new TrieNode();
        for (string& product : products) {
            insert(root, product);
        }

        vector<vector<string>> ans;
        TrieNode* curr = root;
        for (char c : searchWord) {
            c -= 'a';
            if (curr && curr->child[c]) {
                curr = curr->child[c];
                ans.push_back(getWords(curr, 3));
            } else {
                curr = nullptr;
                ans.emplace_back();
            }
        }
        return ans;
    }
};

File: 1268-search-suggestions-system/1268-search-suggestions-system.java
class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        PriorityQueue<String> pq = new PriorityQueue<>(3, (s1, s2) -> s1.compareTo(s2)); 
        List<List<String>> result = new ArrayList<>();
        for (int i = 1; i <= searchWord.length(); i++) {
            String curr = searchWord.substring(0, i);
            for (String product : products) {
                if (product.startsWith(curr)) {
                    pq.offer(product);
                }
            }
            List<String> currResult = new ArrayList<>();
            for (int j = 0; j < 3; j++) {
                if (pq.peek() != null) {
                    currResult.add(pq.poll());
                }
            }
            pq.clear();
            result.add(currResult);
        }
        return result;
    }
}